src/strings/needleman_wunsch/helpers.rs

```diff
@@ -0,0 +1,313 @@
+/// penalty for a match. Our implementation minimizes the total penalty.
+///
+/// # Arguments
+///
+/// * `x`: The first sequence.
+/// * `y`: The second sequence.
+///
+/// # Returns
+///
+/// A nested vector of tuples of total-penalty and Direction, representing the
+/// best alignment at each position.
+pub fn compute_table<U: UInt>(x: &str, y: &str, penalties: Penalties<U>) -> Vec<Vec<(U, Dir)>> {
+    // Initializing table; the inner vectors represent rows in the table.
+    let mut table = vec![vec![(U::zero(), Dir::D); x.len() + 1]; y.len() + 1];
+
+    // The top-left cell starts with a total penalty of zero and no direction.
+    table[0][0] = (U::zero(), Dir::D);
+
+    // Initialize left-most column of distance values.
+    for (i, row) in table.iter_mut().enumerate().skip(1) {
+        row[0] = (penalties.gap * U::from(i), Dir::U);
+    }
+
+    // Initialize top row of distance values.
+    for (j, cell) in table[0].iter_mut().enumerate().skip(1) {
+        *cell = (penalties.gap * U::from(j), Dir::L);
+    }
+
+    // Set values for the body of the table
+    for (i, y_c) in y.chars().enumerate() {
+        for (j, x_c) in x.chars().enumerate() {
+            // Check if sequences match at position `i` in `x` and `j` in `y`.
+            let mismatch_penalty = if x_c == y_c {
+                penalties.match_
+            } else {
+                penalties.mismatch
+            };
+
+            // Compute the three possible penalties and use the minimum to set
+            // the value for the next entry in the table.
+            let d00 = (table[i][j].0 + mismatch_penalty, Dir::D);
+            let d01 = (table[i][j + 1].0 + penalties.gap, Dir::U);
+            let d10 = (table[i + 1][j].0 + penalties.gap, Dir::L);
+
+            table[i + 1][j + 1] = min2(d00, min2(d01, d10));
+        }
+    }
+
+    table
+}
+
+/// Returns the minimum of two penalties, defaulting to the first input.
+fn min2<U: UInt>(a: (U, Dir), b: (U, Dir)) -> (U, Dir) {
+    if a.0 <= b.0 {
+        a
+    } else {
+        b
+    }
+}
+
+/// Converts two aligned sequences into vectors of edits.
+///
+/// # Arguments
+///
+/// * `x`: The first aligned sequence.
+/// * `y`: The second aligned sequence.
+///
+/// # Returns
+///
+/// A 2-slice of Vec<Edit>, each containing the edits needed to convert one
+/// sequence into the other
+pub fn compute_edits(x: &str, y: &str) -> [Vec<Edit>; 2] {
+    [_x_to_y(x, y), _x_to_y(y, x)]
+}
+
+/// Helper for `compute_edits` to compute the edits for `x` into `y`.
+fn _x_to_y(x: &str, y: &str) -> Vec<Edit> {
+    x.chars()
+        .zip(y.chars())
+        .filter(|(x, y)| x != y)
+        .enumerate()
+        .map(|(i, (x, y))| {
+            if y == '-' {
+                Edit::Del(i)
+            } else if x == '-' {
+                Edit::Ins(i, y)
+            } else {
+                Edit::Sub(i, y)
+            }
+        })
+        .collect()
+}
```

Comment:
These still aren't quite right if we want to (easily) reconstruct from these edits. I'm assuming you're okay with that for right now. Unless I missed something, you're still calling this when the sequences are not reversed, so indices of edits after deletion will be incorrect.

Reply:
I guess we can account for that in the reconstruction function though. This would be easier if compute_edits returned the edits and indices based on the reversed aligned sequences